Add `quantiles` aggregation result to grouping

johansolbakken opened the pull request:
Parses `quantiles([float,...], expr)` in grouping parser.

Adds `QuantileAggregator` to public Java API.

Adds `KLLSketch` wrapper in C++.

Java incubator is needed to serialize/deserialize the sketches in datasketches. The flags to include this differ between JDK versions. Adds appropriate flags depending on java version in XML for tests and in Go for runtime.

Moves Json/XML producers to vespajlib since it is required by `QuantileAggregationResult` to produce output as JSON.

@arnej27959 please review
@bjorncs FYI

bjorncs committed:
chore: import `jdk.incubator.foreign` package on JDK17/JDK18

johansolbakken committed:
Add KLLSketch wrapper to vespalib

johansolbakken committed:
Add QuantileAggregationResult to searchlib backend

johansolbakken committed:
Add QuantileAggregationResult to searchlib frontend

johansolbakken committed:
Add incubator flags for Java unit tests

johansolbakken committed:
Add flags to include incubator at runtime

`datasketches-java` for JDK 17 requires `datasketches-memory` library
which uses incubator for serialization.

This commit detects the JDK version and adds appropriate flags for
adding incubator module at runtime.

johansolbakken committed:
Convert QuantileAggregator to QuantielAggregationResult

johansolbakken committed:
Move JsonProducer and XmlProducer to vespajlib

This is required to make a QuantileResult that can implement
JsonProducer.

johansolbakken committed:
Parse quantiles in grouping

arnej27959 commented on searchlib/src/vespa/searchlib/aggregation/quantile_aggregation_result.cpp:
```diff
@@ -0,0 +1,116 @@
+}
+
+void QuantileAggregationResult::onPrepare(const ResultNode& result, bool useForInit) {
+    (void) result;
```
consider logging a warning if useForInit is true here (I'm pretty sure that shouldn't happen)

arnej27959 reviewed (commented):
overall looks great, only minor comments and cosmetic fixes needed as far as I can see.

johansolbakken committed:
Update container-search/src/main/java/com/yahoo/search/grouping/request/QuantileAggregator.java

Co-authored-by: Arne H Juul <arnej27959@users.noreply.github.com>

johansolbakken committed:
Update container-search/src/main/java/com/yahoo/search/grouping/request/QuantileAggregator.java

Co-authored-by: Arne H Juul <arnej27959@users.noreply.github.com>

johansolbakken committed:
Update container-search/src/main/java/com/yahoo/search/grouping/request/QuantileAggregator.java

Co-authored-by: Arne H Juul <arnej27959@users.noreply.github.com>

johansolbakken committed:
Update container-search/src/main/java/com/yahoo/search/grouping/vespa/ResultBuilder.java

Co-authored-by: Arne H Juul <arnej27959@users.noreply.github.com>

johansolbakken commented on searchlib/src/main/java/com/yahoo/searchlib/aggregation/QuantileAggregationResult.java:
```diff
@@ -0,0 +1,203 @@
+
+
+    /**
+     * Constructs an empty result node. <b>NOTE:</b> This instance is broken until non-optional member data is set.
```
The sketch actually throws an exception if it is empty. If we would like it to be a no op, I could map all quantiles to the value zero when the sketch is empty.

johansolbakken committed:
Remove unused constructor and warn when useForInit is true

johansolbakken committed:
Rename `_rank` to `_no_rank`

`onGetRank` requires us to return a result, but this does not apply for
QuantileAggregationResult which contains multiple quantiles. `_no_rank` is only
used to satisfy `onGetRank`.

johansolbakken committed:
Throw exception if quantiles not in range

johansolbakken committed:
Test serialize multiple quantiles

johansolbakken commented:
I have now taken care of the suggestions. Please review and merge @arnej27959

arnej27959 reviewed (approved):
all looks good to me 👍
